### PR TITLE
 [SYCL] List unsupported platforms in sycl-ls verbose mode

### PR DESCRIPTION
**Pull request by jchlanda**
The runtime is able to recognised [banned platforms](https://intel.github.io/llvm-docs/doxygen/namespacesycl_1_1__V1_1_1detail.html#af47bb6df075ff6aa60fc6855d59f6fe6), this patch provides a way to list them using `sycl-ls`' `verbose` switch.

### CODE REVIEW

**Comment by jchlanda** [comment]
This was suggested a while back, here: https://github.com/intel/llvm/pull/6878#issuecomment-1483563291

File: sycl/include/sycl/platform.hpp [hunk 1]
```diff
@@ -194,6 +194,11 @@ class __SYCL_EXPORT platform : public detail::OwnerLessBase<platform> {
   /// \return a vector of all available SYCL platforms.
   static std::vector<platform> get_platforms();
 
+  /// Returns all banned (non-SYCL) platforms in the system.
+  ///
+  /// \return a vector of all banned non-SYCL platforms.
+  static std::vector<platform> get_banned_platforms();
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/platform.hpp:200]: I would rename "banned" to something neutral. "filtered"? something else?

Comment (jchlanda) [reply sycl/include/sycl/platform.hpp:200]: I got inspired by [`IsBannedPlatform`](https://github.com/jchlanda/llvm/blob/jakub/sycl_ls_banned/sycl/source/detail/platform_impl.cpp#L72), but you are right, it does sound quite harsh. Now that I think of it `unsupported` seems right, would that work?

**Review by againull — APPROVED** [review]

File: sycl/source/detail/platform_impl.cpp [hunk 2]
```diff
@@ -96,26 +96,28 @@ static bool IsBannedPlatform(platform Platform) {
...
-    for (const auto &UrPlatform : UrPlatforms) {
-      platform Platform = detail::createSyclObjFromImpl<platform>(
-          getOrMakePlatformImpl(UrPlatform, Plugin));
+  for (const auto &UrPlatform : UrPlatforms) {
+    platform Platform = detail::createSyclObjFromImpl<platform>(
+        getOrMakePlatformImpl(UrPlatform, Plugin));
+    if (!Supported) {
+      if (IsBannedPlatform(Platform)) {
+        Platforms.push_back(Platform);
+      }
```
Comment (al42and) [inline sycl/source/detail/platform_impl.cpp:119]: Currently, we will print supported platforms with available devices, and unsupported/banned platforms with any number of devices. What about supported platforms with no devices?

Comment (jchlanda) [reply sycl/source/detail/platform_impl.cpp:119]: In all honesty I never though of it. It'll be easy enough to add.

Comment (jchlanda) [reply sycl/source/detail/platform_impl.cpp:123]: Done in: 03de71e8e0faeef59db9a38bf8e55b4cc0e7345e

**Comment by jchlanda** [comment]
@dm-vodopyanov @al42and would you like me to do anything else for this patch?

File: sycl/source/detail/platform_impl.cpp (hunk 2 above)
Comment (al42and) [reply sycl/source/detail/platform_impl.cpp:123]: Thanks!

**Comment by dm-vodopyanov** [comment]
> @dm-vodopyanov @al42and would you like me to do anything else for this patch?

Thanks, nothing from my side.

**Comment by jchlanda** [comment]
@intel/llvm-gatekeepers this should be ready to go.

Thank you!

**Comment by dm-vodopyanov** [comment]
> @intel/llvm-gatekeepers this should be ready to go.
> 
> Thank you!

@jchlanda, we need a review from @intel/unified-runtime-reviewers:

image

**Comment by jchlanda** [comment]
> > @intel/llvm-gatekeepers this should be ready to go.
> > Thank you!
> 
> @jchlanda, we need a review from @intel/unified-runtime-reviewers:
> 
> image

My bad sorry, was a bit too hasty with this one.

@intel/unified-runtime-reviewers would you be so kind and take a look at this, thank you.

**Review by aarongreig — APPROVED** [review]

**Comment by jchlanda** [comment]
@intel/llvm-gatekeepers we've got all the approvals now, should be ready to go.

Thank you!

**Comment by jchlanda** [comment]
> @jchlanda - It looks like this made Plugin/sycl-ls.cpp fail on the HIP machine in post-commit. Could you please have a look?

Yeap, on it.

File: sycl/include/sycl/platform.hpp [hunk 3]
```diff
@@ -194,6 +194,11 @@ class __SYCL_EXPORT platform : public detail::OwnerLessBase<platform> {
   /// \return a vector of all available SYCL platforms.
   static std::vector<platform> get_platforms();
 
+  /// Returns all unsupported (non-SYCL) platforms in the system.
+  ///
+  /// \return a vector of all unsupported non-SYCL platforms.
+  static std::vector<platform> get_unsupported_platforms();
```
Comment (steffenlarsen) [inline sycl/include/sycl/platform.hpp:200]: @jchlanda - Sorry for the post-commit review, but I think we need to reconsider the design of this. This function is now a public interface of a SYCL class, but is not part of neither the standard nor an extension. We either need an extension written for this, or we need it hidden, either in the `detail` namespace, by making it `private` or both.

**Pull request merged**